chain/chain/src/flat_storage_resharder.rs

```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
+                return FlatStorageReshardingTaskStatus::Failed;
+            }
+        }
+
+        // Apply deltas and then create the flat storage.
+        let apply_result =
+            self.shard_catchup_apply_deltas(shard_uid, flat_head_block_hash, chain_store);
+        let Ok((num_batches_done, flat_head)) = apply_result else {
+            error!(target: "resharding", ?shard_uid, err = ?apply_result.unwrap_err(), "flat storage shard catchup delta application failed!");
+            return FlatStorageReshardingTaskStatus::Failed;
+        };
+        match self.shard_catchup_finalize_storage(shard_uid, &flat_head) {
+            Ok(_) => {
+                let task_status = FlatStorageReshardingTaskStatus::Successful { num_batches_done };
+                info!(target: "resharding", ?shard_uid, ?task_status, "flat storage shard catchup task finished");
+                // At this point we can trigger the reload of memtries.
+                self.scheduler.send(ReshardingRequest::MemtrieReload { shard_uid });
+                task_status
+            }
+            Err(err) => {
+                error!(target: "resharding", ?shard_uid, ?err, "flat storage shard catchup finalize failed!");
+                FlatStorageReshardingTaskStatus::Failed
+            }
+        }
+    }
+
+    /// Applies flat storage deltas in batches on a shard that is in catchup status.
+    ///
+    /// Returns the number of delta batches applied and the final tip of the flat storage.
+    fn shard_catchup_apply_deltas(
+        &self,
+        shard_uid: ShardUId,
+        mut flat_head_block_hash: CryptoHash,
+        chain_store: &ChainStore,
+    ) -> Result<(usize, Tip), Error> {
+        const CATCH_UP_BLOCKS: u32 = 50;
+
+        let mut num_batches_done: usize = 0;
+
+        loop {
+            let _span = tracing::debug_span!(
+                target: "resharding",
+                "shard_catchup_apply_deltas/batch",
+                ?shard_uid,
+                ?flat_head_block_hash,
+                batch_id = ?num_batches_done)
+            .entered();
+            let chain_final_head = chain_store.final_head()?;
+            let mut merged_changes = FlatStateChanges::default();
+            let store = self.runtime.store().flat_store();
+            let mut store_update = store.store_update();
+
+            // Merge deltas from the next blocks until we reach chain final head.
+            for _ in 0..CATCH_UP_BLOCKS {
+                let height = chain_store.get_block_height(&flat_head_block_hash)?;
+                if height > chain_final_head.height {
+                    let msg = format!("flat head moved too far: head = {flat_head_block_hash}, height = {height}, final block height = {}", chain_final_head.height);
+                    return Err(Error::ReshardingError(msg));
+                }
+                // Stop if we reached chain final head.
+                if flat_head_block_hash == chain_final_head.last_block_hash {
```

Comment:
Is it impossible even in presence of forks? I don't have a test yet so this part is still fuzzy in my head.
(Admittedly I copied this check from catchup of flat storage creation)

Reply:
flat head height in the beginning should always be <= final_head height. If we maintain that constraint, then the flat head height of the children should also be <= final_head height.

Final head is sorta defined as "definitely part of blockchain, without forks"

Reply:
The problem is that during split the first constraint (flat head height <= final_head height) is momentarily broken for children shards :(
Because parent state at its flat head and the parent deltas both end up into the children shards.

Reply:
Update: yes `height > final_head.height` should be impossible with current implementation since the task to catchup is starting only if flat head height <= final head height

Reply:
It should be possible. 

One way is to do the same thing I'm doing now for shard catchup. Basically delay the start of the split task until the resharding hash is final.

Another way would be to call `start_resharding` from resharding manager later, when the resharding block is final. It means storing the last resharding hash in DB and going back every block to see check if that block became final.

I don't have a strong opinion but I feel like the first option is easier because it doesn't require read&write of an extra column/field.

Reply:
Yes, that's what I had suggested [here](https://github.com/near/nearcore/pull/12312#discussion_r1820701423)

I don't think it would be possible to easily integrate approach 2 as we usually call resharding functions on epoch change/shard layout change.